apis/apiextensions/v1beta1/composition_types.go

```diff
@@ -133,14 +206,36 @@ type ReadinessCheck struct {
 	MatchInteger int64 `json:"matchInteger,omitempty"`
 }
 
-// Patch is used to patch the field on the base resource at ToFieldPath
-// after piping the value that is at FromFieldPath of the target resource through
-// transformers.
+// PatchSet defines a named list of patches on a composition.
+type PatchSet struct {
+	ID      string  `json:"id"`
+	Patches []Patch `json:"patches"`
+}
+
+// patchSetMap stores a list of patches from a set, against the set ID.
+type patchSetMap map[string][]Patch
+
+// PatchType is type of the patch to be chosen.
+type PatchType string
+
+// Accepted PatchTypes.
+const (
+	PatchTypeStandard  PatchType = "std"
+	PatchTypeReference PatchType = "ref"
+)
+
+// Patch can either be a reference or a standard patch. Patch type is decided
+// based on the Type field. A reference patch is replaced at reconciliation time
+// with the standard patches defined by a named PatchSet on the parent Composition.
+// A standard patch is field based, patching the field on the base resource at
+// ToFieldPath after piping the value that is at FromFieldPath of the target
+// resource through transformers.
 type Patch struct {
 
 	// FromFieldPath is the path of the field on the upstream resource whose value
 	// to be used as input.
-	FromFieldPath string `json:"fromFieldPath"`
+	// +optional
+	FromFieldPath string `json:"fromFieldPath,omitempty"`
```

Comment:
Ah - I missed this earlier, but API conventions call for this to be a `*string` now that it's optional. Ditto for `Type` and `ID` below.

https://github.com/kubernetes/community/blob/69c5a105d684e8de893f3a7a9765e935410e6ed1/contributors/devel/sig-architecture/api-conventions.md#optional-vs-required